test/instrumentation/index.test.js

```diff
@@ -187,101 +188,6 @@ test('stack branching - no parents', function (t) {
   }, 50)
 })
 
-test('currentTransaction missing - recoverable', function (t) {
-  resetAgent(2, function (data) {
-    t.strictEqual(data.transactions.length, 1)
-    t.strictEqual(data.spans.length, 1)
-    const trans = data.transactions[0]
-    const name = 't0'
-    const span = findObjInArray(data.spans, 'name', name)
-    t.ok(span, 'should have span named ' + name)
-    t.strictEqual(span.transaction_id, trans.id, 'should belong to correct transaction')
-    t.end()
-  })
-  var ins = agent._instrumentation
-  var t0
-
-  var trans = ins.startTransaction('foo')
-  setImmediate(function () {
-    t0 = ins.startSpan('t0')
-    ins.currentTransaction = undefined
-    setImmediate(function () {
-      t0.end()
-      setImmediate(function () {
-        ins.currentTransaction = trans
-        trans.end()
-      })
-    })
-  })
-})
-
-test('currentTransaction missing - not recoverable - last span failed', function (t) {
-  resetAgent(2, function (data) {
-    t.strictEqual(data.transactions.length, 1)
-    t.strictEqual(data.spans.length, 1)
-    const trans = data.transactions[0]
-    const name = 't0'
-    const span = findObjInArray(data.spans, 'name', name)
-    t.ok(span, 'should have span named ' + name)
-    t.strictEqual(span.transaction_id, trans.id, 'should belong to correct transaction')
-    t.end()
-  })
-  var ins = agent._instrumentation
-  var t0, t1
-
-  var trans = ins.startTransaction('foo')
-  setImmediate(function () {
-    t0 = ins.startSpan('t0')
-    setImmediate(function () {
-      t0.end()
-      ins.currentTransaction = undefined
-      t1 = ins.startSpan('t1')
-      t.strictEqual(t1, null)
-      setImmediate(function () {
-        ins.currentTransaction = trans
-        trans.end()
-      })
-    })
-  })
-})
-
-test('currentTransaction missing - not recoverable - middle span failed', function (t) {
-  resetAgent(3, function (data) {
-    t.strictEqual(data.transactions.length, 1)
-    t.strictEqual(data.spans.length, 2)
-    const trans = data.transactions[0]
-    const names = ['t0', 't2']
-    for (const name of names) {
-      const span = findObjInArray(data.spans, 'name', name)
-      t.ok(span, 'should have span named ' + name)
-      t.strictEqual(span.transaction_id, trans.id, 'should belong to correct transaction')
-    }
-    t.end()
-  })
-  var ins = agent._instrumentation
-  var t0, t1, t2
-
-  var trans = ins.startTransaction('foo')
-  setImmediate(function () {
-    t0 = ins.startSpan('t0')
-    setImmediate(function () {
-      ins.currentTransaction = undefined
-      t1 = ins.startSpan('t1')
-      t.strictEqual(t1, null)
-      setImmediate(function () {
-        t0.end()
-        t2 = ins.startSpan('t2')
-        setImmediate(function () {
-          t2.end()
-          setImmediate(function () {
-            trans.end()
-          })
-        })
-      })
-    })
-  })
-})
-
```

Comment:
REVIEW NOTE: This PR drops `Instrumentation#_recoverTransaction(trans)`. It was a function that would reset the currentTransaction to the given one (and log a warning) if they differed. This was called in two places:

1. Originally `_recoverTransaction` and the "recoverable"-named test cases in "test/instrumentation/index.test.js" (that I've now removed) were added in [this commit from 5 years ago](https://github.com/elastic/apm-agent-nodejs/commit/f9d15b55fc469edccdf91878327f8b75a49ff3d1) in relation to dealing with internal user-land callback queues (particularly in db modules). See https://www.youtube.com/watch?v=omOtwqffhck&t=892s which describes the problem. That commit added `Instrumentation#_recoverTransaction(trans)` and called it from what was to become `Span#end()` (after renamings).
    
    A better answer for ensuring these delayed and possibly user-land-queued callbacks are executed in the appropriate run context is to wrap them with `ins.bindFunction(callback)`. This has already been done for any instrumentations that were failing the test suite because of this:
    - memcached: `query.callback = agent._instrumentation.bindFunction(...)`
    - mysql: `return ins.bindFunction(function wrappedCallback ...`
    - tedious: `request.userCallback = ins.bindFunction(...`
    - pg: `args[index] = agent._instrumentation.bindFunction(...`
    - s3: `request.on('complete', ins.bindFunction(...`
    
    The "recoverable" tests here are not meaningfully translatable to the new RunContext management. What is more meaningful are tests on the particular instrumented modules, e.g. see the changes to "test/instrumentation/modules/memcached.test.js". Also each of the memcached, mysql, tedious, and pg tests failed when `_recoverTransaction` was made a no-op and before the added `bindFunction` calls. In other words there were meaningful tests for these cases.

2. Much later `_recoverTransaction` was added in http-shared.js in the "response" handler for outgoing HTTP requests, e.g. calls to `http.request(url, onResponse)`.

    ```js
        function onresponse (res) {
          // Work around async_hooks bug in Node.js 12.0 - 12.2 (https://github.com/nodejs/node/pull/27477)
          ins._recoverTransaction(span.transaction)
          ...
    ```

    With the old/current "async-hooks.js"-based context management and with Node.js v12.0 - 12.2, when this onresponse ran, the `currentTransaction` was null, because of [this core Node async_hooks bug](https://github.com/nodejs/node/pull/27477) where an "init" async hook would not be called for a re-used HTTPParser.

    This was worked around in the agent in #1339 by adding `ins._recoverTransaction(span.transaction)` which just sets `ins.currentTransaction = span.transaction`.  Before the workaround, `node test/instrumentation/modules/http/outgoing.test.js` would fail:

    ```
    not ok 19 test exited without ending: http.request(options, callback)
    ---
      operator: fail
      at: process.<anonymous> (.../node_modules/tape/index.js:94:23)
      stack: |-
        Error: test exited without ending: http.request(options, callback)
    ```

    A possibly better workaround would have been to use ins.bindFunction to bind this `onresponse` to the current span and transaction.

    With the new AsyncHooksRunContextManager, neither of these are necessary. The `init` async hook *is* still missed in node v12.0-12.2, but this `onresponse()` runs within the context of this trans and span in all outgoing.test.js and in manual tests. I.e. adding `assert(ins.currSpan() === span)` here, still passes the test suite (with all versions of node).